// @bgotink/playwright-coverage/lib/data.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.attachmentName = void 0;
exports.getSourceMap = getSourceMap;
exports.getSourceMaps = getSourceMaps;
exports.convertToIstanbulCoverage = convertToIstanbulCoverage;
const fs_1 = require("fs");
const istanbul_lib_coverage_1 = require("istanbul-lib-coverage");
const micromatch_1 = require("micromatch");
const path_1 = require("path");
const url_1 = require("url");
const v8_to_istanbul_1 = __importDefault(require("v8-to-istanbul"));
const convertSourceMap = __importStar(require("convert-source-map"));
exports.attachmentName = '@bgotink/playwright-coverage';
const fetch = import('node-fetch');
async function getSourceMap(url, source) {
    const inlineMap = convertSourceMap.fromSource(source);
    if (inlineMap != null) {
        return inlineMap.sourcemap;
    }
    try {
        const linkedMap = await convertSourceMap.fromMapFileSource(source, async (file) => {
            const resolved = new url_1.URL(file, url);
            switch (resolved.protocol) {
                case 'file:':
                    return await fs_1.promises.readFile(resolved, 'utf8');
                case 'data:': {
                    const comma = resolved.pathname.indexOf(',');
                    const rawData = resolved.pathname.slice(comma + 1);
                    const between = resolved.pathname
                        .slice('application/json;'.length, comma)
                        .split(';');
                    const dataString = between.includes('base64')
                        ? Buffer.from(rawData, 'base64url').toString('utf8')
                        : rawData;
                    return dataString;
                }
                default: {
                    const response = await (await fetch).default(resolved.href, {
                        method: 'GET',
                    });
                    return await response.text();
                }
            }
        });
        if (linkedMap != null) {
            return linkedMap.sourcemap;
        }
    }
    catch {
        return null;
    }
    // No source map comments, try to find an implicit sourcemap at `${url}.map`
    try {
        const response = await (await fetch).default(`${url}.map`, {
            method: 'GET',
        });
        return (await response.json());
    }
    catch {
        return undefined;
    }
}
async function getSourceMaps(sources) {
    return new Map(await Promise.all(Array.from(sources, async ([url, source]) => [url, await getSourceMap(url, source)])));
}
async function convertToIstanbulCoverage(v8Coverage, sources, sourceMaps, exclude, sourceRoot) {
    const istanbulCoverage = (0, istanbul_lib_coverage_1.createCoverageMap)({});
    for (const script of v8Coverage.result) {
        const source = sources.get(script.url);
        const sourceMap = sourceMaps.get(script.url);
        if (source == null || !(sourceMap === null || sourceMap === void 0 ? void 0 : sourceMap.mappings)) {
            continue;
        }
        function sanitizePath(path) {
            let url;
            try {
                url = new url_1.URL(path);
            }
            catch {
                url = (0, url_1.pathToFileURL)(path);
            }
            let relativePath;
            if (url.protocol.startsWith('webpack')) {
                relativePath = url.pathname.slice(1); // webpack: URLs contain relative paths
            }
            else {
                relativePath = url.pathname;
            }
            if (relativePath.includes('/webpack:/')) {
                // v8ToIstanbul breaks when the source root in the source map is set to webpack:
                // It treats the URL as a path, leading to a confusing result.
                relativePath = relativePath.slice(relativePath.indexOf('/webpack:/') + '/webpack:/'.length);
            }
            else if (path_1.posix.isAbsolute(relativePath)) {
                relativePath = path_1.posix.relative((0, url_1.pathToFileURL)(sourceRoot).pathname, path);
            }
            return relativePath;
        }
        const isExcludedCache = new Map();
        const convertor = (0, v8_to_istanbul_1.default)(
        // This path is used to resolve files, but the filename doesn't matter
        (0, path_1.join)(sourceRoot, 'unused.js'), 0, (sourceMap === null || sourceMap === void 0 ? void 0 : sourceMap.mappings)
            ? {
                source,
                sourceMap: { sourcemap: sourceMap },
            }
            : {
                source: convertSourceMap.removeMapFileComments(convertSourceMap.removeComments(source)),
            }, path => {
            let isExcluded = isExcludedCache.get(path);
            if (isExcluded != null) {
                return isExcluded;
            }
            const relativePath = sanitizePath(path);
            isExcluded =
                // ignore files outside of the root
                relativePath.startsWith('../') ||
                    // ignore webpack files
                    path.includes('/webpack:/webpack/') ||
                    relativePath === 'webpack/bootstrap' ||
                    relativePath.startsWith('webpack/runtime/') ||
                    // ignore dependencies
                    relativePath.startsWith('node_modules/') ||
                    relativePath.includes('/node_modules/') ||
                    // apply exclusions
                    (0, micromatch_1.isMatch)(relativePath, exclude);
            isExcludedCache.set(path, isExcluded);
            return isExcluded;
        });
        try {
            await convertor.load();
        }
        catch (error) {
            continue;
        }
        convertor.applyCoverage(script.functions);
        istanbulCoverage.merge(Object.fromEntries(Array.from(Object.entries(convertor.toIstanbul()), ([path, coverage]) => {
            path = sanitizePath(path);
            return [
                path,
                {
                    ...coverage,
                    path,
                },
            ];
        })));
    }
    return istanbulCoverage;
}
