// @bgotink/playwright-coverage/lib/fixtures.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mixinFixtures = mixinFixtures;
const fs_1 = require("fs");
const data_1 = require("./data");
const coverageFixtures = {
    collectCoverage: [true, { option: true }],
    page: async ({ page, collectCoverage, browserName }, use, testInfo) => {
        if (browserName !== 'chromium' || !collectCoverage) {
            return use(page);
        }
        await page.coverage.startJSCoverage({
            resetOnNavigation: false,
        });
        await use(page);
        const result = await page.coverage.stopJSCoverage();
        const resultFile = testInfo.outputPath('v8-coverage.json');
        await fs_1.promises.writeFile(resultFile, JSON.stringify({ result }));
        testInfo.attachments.push({
            name: data_1.attachmentName,
            contentType: 'application/json',
            path: resultFile,
        });
    },
};
function mixinFixtures(base) {
    return base.extend(coverageFixtures);
}
